File: backend/app/api/guias_despacho.py
from datetime import date, datetime, timezone
from decimal import Decimal
from fastapi import APIRouter, HTTPException, Query, status
from fastapi.responses import Response
from sqlalchemy.orm import Session, joinedload, selectinload

from app.api.deps import require_permission
from app.api.dte import _next_numero
from app.models.dte_emision import DteEmision
from app.models.empresa import Empresa
from app.models.guia_despacho import GuiaDespacho, GuiaDespachoLinea
from app.models.system_config import SystemConfig
from app.models.user import User
from app.schemas.guia_despacho import (
    GuiaDespachoCreate,
    GuiaDespachoListOut,
    GuiaDespachoOut,
    GuiaDespachoUpdate,
    GuiaEmailBody,
)
from app.services.email import EmailNotConfiguredError, enviar_guia_despacho as _enviar_guia_email
from app.services.pdf import generar_pdf_guia_despacho
from app.utils.logo import empresa_logo_data_uri
from app.tasks.dte import emit_dte

router = APIRouter()


def _config_dict(db: Session) -> dict:
    return {r.key: r.value for r in db.query(SystemConfig).all()}


def _calcular_lineas_y_totales_guia(guia: GuiaDespacho) -> None:
    """En guía, precio_unitario es BRUTO (con IVA), igual que boleta."""
    total_neto = Decimal("0")
    total_iva = Decimal("0")
    total_bruto = Decimal("0")
    tasa = Decimal("0.19")

    for linea in guia.lineas:
        bruto_unit = linea.precio_unitario
        cantidad = linea.cantidad
        descuento = linea.descuento_pct or Decimal("0")
        bruto = bruto_unit * cantidad * (Decimal("1") - descuento / Decimal("100"))

        if linea.exenta:
            neto = bruto.quantize(Decimal("0.01"))
            iva_linea = Decimal("0")
        else:
            neto = (bruto / (Decimal("1") + tasa)).quantize(Decimal("0.01"))
            iva_linea = (bruto - neto).quantize(Decimal("0.01"))

        linea.total_neto = neto
        linea.iva = iva_linea
        linea.total_linea = (neto + iva_linea).quantize(Decimal("0.01"))
        total_neto += linea.total_neto
        total_iva += linea.iva
        total_bruto += linea.total_linea

    guia.total_neto = total_neto
    guia.total_iva = total_iva
    guia.total = total_bruto


def _load_guia(db: Session, guia_id: int) -> GuiaDespacho:
    guia = (
        db.query(GuiaDespacho)
        .options(
            joinedload(GuiaDespacho.cliente),
            joinedload(GuiaDespacho.vendedor),
            selectinload(GuiaDespacho.lineas),
        )
        .filter(GuiaDespacho.id == guia_id)
        .first()
    )
    if not guia:
        raise HTTPException(status_code=404, detail="Guía no encontrada")
    return guia


@router.post("/", response_model=GuiaDespachoOut, status_code=status.HTTP_201_CREATED)
def crear_guia_despacho(
    body: GuiaDespachoCreate,
    perms: tuple[User, Session] = require_permission("guias_despacho", "create"),
):
    current_user, db = perms
    try:
        numero = _next_numero(db, "guia_despacho_last_id")
        guia = GuiaDespacho(
            numero=numero,
            fecha=body.fecha or date.today(),
            motivo_traslado=body.motivo_traslado,
            direccion_destino=body.direccion_destino,
            comuna_destino=body.comuna_destino,
            cliente_id=body.cliente_id,
            empresa_id=body.empresa_id,
            nota_venta_id=body.nota_venta_id,
            email_envio=body.email_envio,
            vendedor_id=current_user.id,
        )
        db.add(guia)
        db.flush()

        for l in body.lineas:
            db.add(GuiaDespachoLinea(
                guia_despacho_id=guia.id,
                orden=l.orden,
                producto_id=l.producto_id,
                descripcion=l.descripcion,
                cantidad=l.cantidad,
                precio_unitario=l.precio_unitario,
                descuento_pct=l.descuento_pct,
                exenta=l.exenta,
            ))
        db.flush()
        db.refresh(guia)
        _calcular_lineas_y_totales_guia(guia)

        # Guía DTE 52 NO descuenta stock — el documento tributario asociado lo hace
        # (ver INV-04 / docs/architecture.md). Invariante intencional hasta Phase 3 (D-13).

        emision = DteEmision(
            tipo="052",  # 3 chars con cero líder (Pitfall 8)
            guia_despacho_id=guia.id,
            monto_neto=int(guia.total_neto),
            monto_iva=int(guia.total_iva),
            monto_total=int(guia.total),
        )
        db.add(emision)
        db.flush()

        guia.dte_estado = "pendiente"
        db.commit()
        db.refresh(guia)
    except Exception:
        db.rollback()
        raise

    emit_dte.delay(emision.id)
    return guia


@router.get("/", response_model=list[GuiaDespachoListOut])
def listar_guias_despacho(
    estado: str | None = Query(None),
    dte_estado: str | None = Query(None),
    cliente_id: int | None = Query(None),
    desde: date | None = Query(None),
    hasta: date | None = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, le=200),
    perms: tuple[User, Session] = require_permission("guias_despacho", "view"),
):
    _, db = perms
    q = db.query(GuiaDespacho)
    if estado:
        q = q.filter(GuiaDespacho.estado == estado)
    if dte_estado:
        q = q.filter(GuiaDespacho.dte_estado == dte_estado)
    if cliente_id:
        q = q.filter(GuiaDespacho.cliente_id == cliente_id)
    if desde:
        q = q.filter(GuiaDespacho.fecha >= desde)
    if hasta:
        q = q.filter(GuiaDespacho.fecha <= hasta)
    return q.order_by(GuiaDespacho.id.desc()).offset(skip).limit(limit).all()


@router.get("/{guia_id}", response_model=GuiaDespachoOut)
def obtener_guia(
    guia_id: int,
    perms: tuple[User, Session] = require_permission("guias_despacho", "view"),
):
    _, db = perms
    return _load_guia(db, guia_id)


@router.patch("/{guia_id}", response_model=GuiaDespachoOut)
def editar_guia(
    guia_id: int,
    body: GuiaDespachoUpdate,
    perms: tuple[User, Session] = require_permission("guias_despacho", "edit"),
):
    _, db = perms
    guia = _load_guia(db, guia_id)
    try:
        for field, value in body.model_dump(exclude_unset=True).items():
            setattr(guia, field, value)
        db.commit()
        db.refresh(guia)
    except Exception:
        db.rollback()
        raise
    return guia


@router.delete("/{guia_id}", status_code=status.HTTP_204_NO_CONTENT)
def eliminar_guia(
    guia_id: int,
    perms: tuple[User, Session] = require_permission("guias_despacho", "delete"),
):
    _, db = perms
    guia = _load_guia(db, guia_id)
    if guia.dte_estado != "no_emitida":
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="No se puede eliminar guía emitida; usá NC para anular",
        )
    try:
        db.delete(guia)
        db.commit()
    except Exception:
        db.rollback()
        raise


@router.get("/export/excel")
def exportar_excel(
    perms: tuple[User, Session] = require_permission("guias_despacho", "view"),
    fecha_desde: str | None = None,
    fecha_hasta: str | None = None,
    cliente_id: int | None = None,
    estado: list[str] | None = Query(None),
    dte_estado: list[str] | None = Query(None),
    motivo_traslado: int | None = None,
    vendedor_id: int | None = None,
):
    user, db = perms
    q = db.query(GuiaDespacho).options(
        joinedload(GuiaDespacho.cliente),
        joinedload(GuiaDespacho.vendedor),
    )
    if user.role == "vendedor":
        q = q.filter(GuiaDespacho.vendedor_id == user.id)
    if fecha_desde:
        q = q.filter(GuiaDespacho.fecha >= fecha_desde)
    if fecha_hasta:
        q = q.filter(GuiaDespacho.fecha <= fecha_hasta)
    if cliente_id:
        q = q.filter(GuiaDespacho.cliente_id == cliente_id)
    if estado:
        q = q.filter(GuiaDespacho.estado.in_(estado))
    if dte_estado:
        q = q.filter(GuiaDespacho.dte_estado.in_(dte_estado))
    if motivo_traslado:
        q = q.filter(GuiaDespacho.motivo_traslado == motivo_traslado)
    if vendedor_id:
        q = q.filter(GuiaDespacho.vendedor_id == vendedor_id)

    rows = q.order_by(GuiaDespacho.fecha.desc(), GuiaDespacho.numero.desc()).all()

    from openpyxl import Workbook
    from io import BytesIO
    wb = Workbook()
    ws = wb.active
    ws.title = "Guías de Despacho"
    ws.append([
        "N°", "Fecha", "Cliente", "RUT", "Motivo", "NV vinculada",
        "Total neto", "IVA", "Total", "Estado", "DTE", "Vendedor",
    ])
    for g in rows:
        ws.append([
            g.numero,
            g.fecha.isoformat() if g.fecha else "",
            g.cliente.nombre if g.cliente else "",
            g.cliente.rut if g.cliente else "",
            g.motivo_traslado,
            g.nota_venta_id or "",
            float(g.total_neto or 0),
            float(g.total_iva or 0),
            float(g.total or 0),
            g.estado,
            g.dte_estado,
            g.vendedor.name if g.vendedor else "",
        ])
    buf = BytesIO()
    wb.save(buf)
    buf.seek(0)
    return Response(
        content=buf.read(),
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": 'attachment; filename="guias-despacho.xlsx"'},
    )


@router.get("/{guia_id}/pdf")
def descargar_pdf_guia(
    guia_id: int,
    perms: tuple[User, Session] = require_permission("guias_despacho", "view"),
):
    _, db = perms
    guia = _load_guia(db, guia_id)
    config = _config_dict(db)
    if guia.empresa_id:
        empresa = db.get(Empresa, guia.empresa_id)
        if empresa:
            uri = empresa_logo_data_uri(empresa.logo_path)
            if uri:
                config["empresa_logo_url"] = uri
    pdf_bytes = generar_pdf_guia_despacho(guia, config)
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={
            "Content-Disposition": f'inline; filename="guia-{guia.numero}.pdf"',
        },
    )


@router.post("/{guia_id}/email", response_model=GuiaDespachoOut)
def enviar_email_guia(
    guia_id: int,
    body: GuiaEmailBody,
    perms: tuple[User, Session] = require_permission("guias_despacho", "edit"),
):
    _, db = perms
    guia = _load_guia(db, guia_id)
    destino = (body.email or guia.email_envio or
               (guia.cliente.email if guia.cliente and getattr(guia.cliente, "email", None) else None))
    if not destino:
        raise HTTPException(status_code=422, detail="No hay email destino")
    config = _config_dict(db)
    if guia.empresa_id:
        empresa = db.get(Empresa, guia.empresa_id)
        if empresa:
            uri = empresa_logo_data_uri(empresa.logo_path)
            if uri:
                config["empresa_logo_url"] = uri
    pdf_bytes = generar_pdf_guia_despacho(guia, config)
    try:
        _enviar_guia_email(guia, pdf_bytes, destino)
    except EmailNotConfiguredError:
        raise HTTPException(status_code=503, detail="Email no configurado")
    try:
        guia.email_enviado_at = datetime.now(timezone.utc)
        db.commit()
        db.refresh(guia)
    except Exception:
        db.rollback()
        raise
    return guia

File: frontend/src/api/guiasDespacho.ts
// frontend/src/api/guiasDespacho.ts
import { api } from '../lib/api';

export type GuiaEstado = 'emitida' | 'anulada';
export type GuiaDteEstado =
  | 'no_emitida'
  | 'pendiente'
  | 'procesando'
  | 'aceptada'
  | 'rechazada';

export type MotivoTraslado = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

export const MOTIVOS_TRASLADO: { value: MotivoTraslado; label: string }[] = [
  { value: 1, label: '1 — Operación constituye venta' },
  { value: 2, label: '2 — Ventas por entregar' },
  { value: 3, label: '3 — Consignaciones' },
  { value: 4, label: '4 — Entrega gratuita' },
  { value: 5, label: '5 — Traslado interno' },
  { value: 6, label: '6 — Otros traslados no venta' },
  { value: 7, label: '7 — Guía de devolución' },
  { value: 8, label: '8 — Traslado para exportación' },
  { value: 9, label: '9 — Venta para exportación' },
];

export interface ClienteMin {
  id: number;
  nombre: string;
  rut?: string | null;
}

export interface VendedorMin {
  id: number;
  name: string;
}

export interface GuiaLineaInput {
  orden?: number;
  producto_id?: number | null;
  descripcion: string;
  cantidad: string;
  precio_unitario: string;
  descuento_pct?: string;
  exenta?: boolean;
}

export interface GuiaLinea {
  id: number;
  orden: number;
  producto_id?: number | null;
  descripcion: string;
  cantidad: string;
  precio_unitario: string;
  descuento_pct: string;
  exenta: boolean;
  total_neto: string;
  iva: string;
  total_linea: string;
}

export interface GuiaDespacho {
  id: number;
  numero: number;
  fecha: string;
  cliente_id?: number | null;
  empresa_id?: number | null;
  nota_venta_id?: number | null;
  motivo_traslado: MotivoTraslado;
  direccion_destino: string;
  comuna_destino: string;
  email_envio?: string | null;
  vendedor_id?: number | null;
  total_neto: string;
  total_iva: string;
  total: string;
  estado: GuiaEstado;
  dte_estado: GuiaDteEstado;
  folio_sii?: number | null;
  track_id?: string | null;
  email_enviado_at?: string | null;
  created_at: string;
  updated_at: string;
  cliente?: ClienteMin | null;
  vendedor?: VendedorMin | null;
  nota_venta?: { id: number; numero: number } | null;
  lineas: GuiaLinea[];
}

export interface GuiaDespachoListItem {
  id: number;
  numero: number;
  fecha: string;
  cliente_id?: number | null;
  motivo_traslado: MotivoTraslado;
  nota_venta_id?: number | null;
  total: string;
  estado: GuiaEstado;
  dte_estado: GuiaDteEstado;
  cliente?: ClienteMin | null;
  vendedor?: VendedorMin | null;
}

export interface GuiaCreatePayload {
  fecha?: string;
  cliente_id: number;
  empresa_id?: number | null;
  nota_venta_id?: number | null;
  motivo_traslado: MotivoTraslado;
  direccion_destino: string;
  comuna_destino: string;
  email_envio?: string;
  lineas: GuiaLineaInput[];
}

export interface GuiaPatchPayload {
  direccion_destino?: string;
  comuna_destino?: string;
  email_envio?: string | null;
}

export interface GuiaListFilters {
  fecha_desde?: string;
  fecha_hasta?: string;
  cliente_id?: number;
  empresa_id?: number;
  motivo_traslado?: MotivoTraslado;
  estado?: GuiaEstado[];
  dte_estado?: GuiaDteEstado[];
  vendedor_id?: number;
  limit?: number;
  offset?: number;
}

export interface GuiaDespachoListResponse {
  data: GuiaDespachoListItem[];
  pagination: { limit: number; offset: number; total: number };
}

function cleanParams(filtros: GuiaListFilters): Record<string, unknown> {
  return Object.fromEntries(
    Object.entries(filtros).filter(([, v]) => {
      if (v == null) return false;
      if (typeof v === 'string' && v === '') return false;
      if (Array.isArray(v) && v.length === 0) return false;
      return true;
    }),
  );
}

export async function listarGuiasDespacho(
  filtros: GuiaListFilters = {},
): Promise<GuiaDespachoListResponse> {
  const params = cleanParams(filtros);
  const { data } = await api.get<GuiaDespachoListResponse>('/api/guias-despacho/', { params });
  return data;
}

export async function getGuiaDespacho(id: number): Promise<GuiaDespacho> {
  const { data } = await api.get<GuiaDespacho>(`/api/guias-despacho/${id}`);
  return data;
}

export async function crearGuiaDespacho(payload: GuiaCreatePayload): Promise<GuiaDespacho> {
  const { data } = await api.post<GuiaDespacho>('/api/guias-despacho/', payload);
  return data;
}

export async function patchGuiaDespacho(
  id: number,
  payload: GuiaPatchPayload,
): Promise<GuiaDespacho> {
  const { data } = await api.patch<GuiaDespacho>(`/api/guias-despacho/${id}`, payload);
  return data;
}

export async function eliminarGuiaDespacho(id: number): Promise<void> {
  await api.delete(`/api/guias-despacho/${id}`);
}

export async function emitirGuiaDespachoDte(id: number): Promise<GuiaDespacho> {
  const { data } = await api.post<GuiaDespacho>(`/api/dte/guias-despacho/${id}/emitir`);
  return data;
}

export async function enviarEmailGuiaDespacho(
  id: number,
  email?: string,
): Promise<GuiaDespacho> {
  const { data } = await api.post<GuiaDespacho>(`/api/guias-despacho/${id}/email`, { email });
  return data;
}

export async function descargarPdfGuiaDespacho(id: number): Promise<Blob> {
  const { data } = await api.get<Blob>(`/api/guias-despacho/${id}/pdf`, {
    responseType: 'blob',
  });
  return data;
}

export async function exportarGuiasDespachoExcel(
  filtros: GuiaListFilters = {},
): Promise<Blob> {
  const params = cleanParams(filtros);
  const { data } = await api.get<Blob>('/api/guias-despacho/export/excel', {
    params,
    responseType: 'blob',
  });
  return data;
}

export function pdfGuiaDespachoUrl(id: number): string {
  return `/api/guias-despacho/${id}/pdf`;
}
